addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java

```diff
@@ -0,0 +1,157 @@
+public class DSMRHandlerFactory extends BaseThingHandlerFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(DSMRHandlerFactory.class);
+
+    private final Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
+    private @NonNullByDefault({}) LocaleProvider localeProvider;
+
+    private @NonNullByDefault({}) TranslationProvider i18nProvider;
+
+    /**
+     * Returns if the specified ThingTypeUID is supported by this handler.
+     *
+     * This handler support the THING_TYPE_DSMR_BRIDGE type and all ThingTypesUID that
+     * belongs to the supported DSMRMeterType objects
+     *
+     * @param thingTypeUID {@link ThingTypeUID} to check
+     * @return true if the specified ThingTypeUID is supported, false otherwise
+     */
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        if (thingTypeUID.equals(DSMRBindingConstants.THING_TYPE_DSMR_BRIDGE)) {
+            logger.debug("DSMR Bridge Thing {} supported", thingTypeUID);
+            return true;
+        } else {
+            boolean thingTypeUIDIsMeter = DSMRMeterType.METER_THING_TYPES.contains(thingTypeUID);
+
+            if (thingTypeUIDIsMeter) {
+                logger.trace("{} is a supported DSMR Meter thing", thingTypeUID);
+            }
+            return thingTypeUIDIsMeter;
+        }
+    }
+
+    /**
+     * Create the ThingHandler for the corresponding Thing
+     *
+     * There are two handlers supported:
+     * - DSMRBridgeHandler that handle the Thing that corresponds to the physical DSMR device and does the serial
+     * communication
+     * - MeterHandler that handles the Meter things that are a logical part of the physical device
+     *
+     * @param thing The Thing to create a ThingHandler for
+     * @return ThingHandler for the given Thing or null if the Thing is not supported
+     */
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        logger.debug("Searching for thingTypeUID {}", thingTypeUID);
+
+        if (DSMRBindingConstants.THING_TYPE_DSMR_BRIDGE.equals(thingTypeUID)) {
+            DSMRBridgeHandler handler = new DSMRBridgeHandler((Bridge) thing);
+            registerDiscoveryService(handler);
+            return handler;
+        } else if (DSMRMeterType.METER_THING_TYPES.contains(thingTypeUID)) {
+            return new DSMRMeterHandler(thing);
+        }
+
+        return null;
+    }
+
+    /**
+     * Registers a meter discovery service for the bridge handler.
+     *
+     * @param bridgeHandler handler to register service for
+     */
+    private synchronized void registerDiscoveryService(DSMRBridgeHandler bridgeHandler) {
+        DSMRMeterDiscoveryService discoveryService = new DSMRMeterDiscoveryService(bridgeHandler);
+
+        discoveryService.setLocaleProvider(localeProvider);
```

Comment:
Hm, if I am not missing anything, neither the `localeProvider` nor the `i18nProvider` are actually used by the discovery service (and normally, there should also be no need for them). So you might want to remove the dependency to these services from the code.

Reply:
The're used by the parent class `AbstractDiscoveryService` to set the label of a newly detected thing. These labels are i18n supported.

Reply:
Oh, thanks, I wasn't even aware of that :-)